pandas/core/algorithms.py

```diff
@@ -812,6 +801,17 @@ def factorize(
             na_sentinel_arg = None
         else:
             na_sentinel_arg = na_sentinel
+
+        if not dropna and not sort:
+            # factorize can now handle differentiating various types of null values.
+            # However, for backwards compatibility we only use the null for the
+            # provided dtype. This may be revisited in the future, see GH#48476.
+            null_mask = isna(values)
```

Comment:
Does it make sense to check for integer dtype? Not sure if this offers any performance gains

Reply:
Is there a way to determine which dtypes can hold nan? I didn't see one in `core.dtypes`, but might have missed it.

For integer dtypes, this op takes 13.1 µs in the ASVs I commented on in the linked PR. There could be some gains here if we can eliminate certain dtypes.

Reply:
Hm, this is probably checked under the hood then. I don't think that this offers much gains then.

The more interesting question is probably, if there is a dtype that can hold different nans apart from object. I don't think so. So we could only check for object dtype?

Reply:
Agreed - we can check only object dtype here. ASVs for this are currently running, I plan update this PR later this evening (GMT-4).